index.html

```diff
@@ -1323,40 +1376,66 @@ <h2>
-              time period in which those calls were made.
+              abuse of the <a>hasEnrolledInstrument()</a> method for
+              fingerprinting purposes, such as creating <a>PaymentRequest</a>
+              objects with a variety of supported <a>payment methods</a> and
+              calling <a>hasEnrolledInstrument()</a> on them one after the
+              other. For example, a user agent may restrict the number of
+              successful calls that can be made based on the <a>top-level
+              browsing context</a> or the time period in which those calls were
+              made.
             </p>
           </li>
-          <li>Let <var>hasHandlerPromise</var> be <a>a new promise</a>.
+          <li>Let <var>promise</var> be <a>a new promise</a>.
           </li>
-          <li>Return <var>hasHandlerPromise</var>, and perform the remaining
-          steps <a>in parallel</a>.
+          <li>Return <var>promise</var> and perform the remaining steps <a>in
+          parallel</a>.
           </li>
           <li>For each <var>paymentMethod</var> tuple in
           <var>request</var>.<a>[[\serializedMethodData]]</a>:
             <ol>
               <li>Let <var>identifier</var> be the first element in the
               <var>paymentMethod</var> tuple.
               </li>
-              <li>If there user agent has a <a>payment handler</a> that support
-              handling payment requests for <var>identifier</var>, resolve
-              <var>hasHandlerPromise</var> with true and terminate this
+              <li>Let <var>data</var> be the result of <a data-cite=
+              "ECMASCRIPT#sec-json.parse">JSON-parsing</a> the second element
+              in the <var>paymentMethod</var> tuple.
+              </li>
+              <li>If required by the specification that defines the
+              <var>identifier</var>, then <a data-cite=
+              "WEBIDL#dfn-convert-ecmascript-to-idl-value">convert</a>
+              <var>data</var> to an IDL value. Otherwise, <a data-cite=
+              "WEBIDL#dfn-convert-ecmascript-to-idl-value">convert</a> to
+              <a data-cite="WEBIDL#idl-object">object</a>.
+              </li>
+              <li>If conversion results in an <a data-cite=
+              "WEBIDL#dfn-exception">exception</a> <var>error</var>, reject
+              <var>promise</var> with <var>error</var> and terminate this
               algorithm.
               </li>
+              <li>Let <var>handlers</var> be a <a>list</a> of registered
+              <a>payment handlers</a> that are authorized and can handle
+              payment request for <var>identifier</var>.
+              </li>
+              <li>For each <var>handler</var> in <var>handlers</var>:
+                <ol>
+                  <li>Let <var>hasEnrolledInstrument</var> be the result of
+                  running <var>handler</var>'s <a>steps to check if a payment
```

Comment:
I have a question here: Instead of "steps to check if a payment can be made" should this be "steps to check if the handler has an enrolled instrument"? And does this mean that a payment method also needs to define those steps? Does a revised basic card spec need to change the steps to be about enrolled instrument?

(I don't have answers to these questions yet.)

[1] https://www.w3.org/TR/payment-method-basic-card/#steps-to-check-if-a-payment-can-be-made

Reply:
In Chrome we're using the existing steps in the basic card spec to determine if a card should be considered "enrolled". For payment handler, I believe the equivalent is the CanMakePaymentEvent. In both cases, our rationale is that the new canMakePayment() method checks the existence of a payment handler for the specified method, and hasEnrolledInstrument() checks the actual "can-make-payment" property of the payment handler.

I think renaming "steps to check if a payment can be made" to "steps to check if the handler has an enrolled instrument" may be more clear in the long run, but it introduces a lot of changes to existing spec without changing any behavior. That's why I opted to reuse the existing language as much as possible.

Reply:
Hi @ianbjacobs - did I answer your question? This is the last open review question blocking this PR. :-)

Reply:
@danyao, sorry for the delay. Thank you for the answer. I'm ok with your proposal.